### PR TITLE
Update template

### PR DESCRIPTION
**Pull request by AlexWayfer**
There are many changes from my experience and notes.

### CODE REVIEW

**Comment by AlexWayfer** [comment]
I was also thinking about README template, but I'd add it after real-life gem experience.

**Comment by olleolleolle** [comment]
Thanks for taking the time to update so many little details. The editor config is gold for people who use editors that support it (nearly everybody), formatting of little things, many nuggets here. Thanks! 

I made minor notes during my reading commit by commit, but those were not blockers to merging. 

I'll let @iMacTia read, too, and then we can move forward.

File: .gitignore [hunk 1]
```diff
@@ -13,5 +13,6 @@ gemfile.lock
 
 .rvmrc
 .ruby-version
+Gemfile.lock
```
Comment (iMacTia) [inline .gitignore]: nit: I can't test this now but I see a lowercase `gemfile.lock` a few lines up.
Is that a bug from the bundler generator? Are we sure we need both?

File: faraday-my_middleware.gemspec [hunk 2]
```diff
@@ -1,8 +1,6 @@
 # frozen_string_literal: true
 
-$LOAD_PATH.push File.expand_path('lib', __dir__)
-
-require 'faraday/my_middleware/version'
+require_relative 'lib/faraday/my_middleware/version'
```
Comment (iMacTia) [inline faraday-my_middleware.gemspec:3]: It's quite a common practice to expand `$LOAD_PATH` so you don't have to use `require_relative`.
I'm not against this change, but curious on why you think this is a better approach?

**Review by iMacTia — COMMENTED** [review]
Thanks for taking the time to provide feedback and propose changes based on your experience with the template repository! I'm sure many other developers will be grateful when they'll use it in future 🙏!

I've left a few comments and questions, but nothing major

File: .gitignore (hunk 1 above)
Comment (olleolleolle) [reply .gitignore]: If we are ignoring the lockfile, then `gems.rb` is another name which Bundler supports, and can be told to create.

Comment (AlexWayfer) [reply .gitignore]: > nit: I can't test this now but I see a lowercase `gemfile.lock` a few lines up.
> Is that a bug from the bundler generator? Are we sure we need both?

Didn't notice that, thank you.

BTW, gitignore is case sensitive by default, although it can be configured: https://stackoverflow.com/q/11116636/2630849

But I'd stick with default for people. 



> If we are ignoring the lockfile, then `gems.rb` is another name which Bundler supports, and can be told to create.

I didn't face with such file (it hasn't became default for Bundler 2), but isn't it an alternative for `Gemfile`, not a lock-file? I don't know a lock-file name for `gems.rb`.

Comment (AlexWayfer) [reply .gitignore]: BTW I changed the commit and just "fixed" casing for `Gemfile.lock`, thank you.

File: faraday-my_middleware.gemspec (hunk 2 above)
Comment (AlexWayfer) [reply faraday-my_middleware.gemspec:3]: > It's quite a common practice to expand `$LOAD_PATH`

I'm not a fan of "use a common practice just because it's common". 😅 (I'm also using tabs indentation even for Ruby 😶 )



> so you don't have to use `require_relative`

I don't see a problem here.



> I'm not against this change, but curious on why you think this is a better approach?

But I see a problem with `require`, and even faced with it. You know, the worst kind of errors, when there is no exception, and everything works, but slightly in another way…

So, if you have a gem installed globally (just `gem install my-gem`) — this `require` becomes ambiguous. And in the old code there is even `push`, not `unshift`, so I'm even more afraid what can happen here.

In short words: `require_relative` is explicit, reliable, doesn't require a global state change (doesn't affect anything), not too longer to write.

**Review by iMacTia — APPROVED** [review]
This is good to go for me, no need to add more as we can always open a new PR later if needed. @olleolleolle do you want to have a final look?

**Comment by AlexWayfer** [comment]
(still don't see "Discussions" section)

You know, we can change it to something more user-friendly, like gem generator.

I'm a developer of (unknown) [`gem_generator`](https://github.com/AlexWayfer/gem_generator), and it's pretty handful, instead of manual check of all file names and code inclusions in different cases (`PascalCase`, `snake_case`, etc.).

Maybe I could add a functionality like template support.

Just a single command (with correct argument to your template) instead of commits [like this (with manual work)](https://github.com/AlexWayfer/faraday-encode_xml/commit/1f17a70ab6b25b53497a4d7bdc4a1a2026702276) (I haven't done the actual README yet).

**Comment by AlexWayfer** [comment]
Also I have a pretty important question:

You have `on_request` and `on_complete` in template, but I see just `call` in [the middleware writing documentation](https://lostisland.github.io/faraday/middleware/custom).

What is correct?

**Comment by iMacTia** [comment]
> You know, we can change it to something more user-friendly, like gem generator

That would be fantastic. At first I was actually looking for a tool that would allow me to define some "template variables" in the gem code and then filling them using variables, but I couldn't find anything that was easy enough to use or setup.
If you know such a tool or are interested to automate this process in `gem_generator`, then I'd definitely be interested.

> Also I have a pretty important question:
> You have on_request and on_complete in template, but I see just call in the middleware writing documentation.
> What is correct?

Both are valid options, `on_request` and `on_complete` are preferred for simple middleware and are mentioned in the same documentation page under the [`Faraday::Middleware` section](https://lostisland.github.io/faraday/middleware/custom#faradaymiddleware).

**Comment by iMacTia** [comment]
Merging this in, thanks again a lot for the great feedback @AlexWayfer 🙏 
I hope your middleware is coming together nicely and feel free to reach out if we can help!
